lib/t/calls.ex

```diff
@@ -292,4 +292,75 @@ defmodule T.Calls do
     |> select([c, p], {c, p})
     |> Repo.all()
   end
+
+  alias T.Media
+
+  def voicemail_s3_url do
+    Media.user_s3_url()
+  end
+
+  def voicemail_url(s3_key) do
+    Media.user_presigned_url(s3_key)
+  end
+
+  def voicemail_upload_form(content_type) do
+    Media.sign_form_upload(
+      key: Ecto.UUID.generate(),
+      content_type: content_type,
+      # 50 MB
+      max_file_size: 50_000_000,
+      expires_in: :timer.hours(1),
+      acl: "private"
+    )
+  end
+
+  @spec voicemail_save_message(uuid, uuid, uuid) ::
+          {:ok, %Voicemail{}} | {:error, reason :: String.t()}
+  def voicemail_save_message(caller_id, match_id, s3_key) do
+    with :ok <- ensure_voicemail_allowed(caller_id, match_id) do
+      voicemail = %Voicemail{caller_id: caller_id, match_id: match_id, s3_key: s3_key}
+
+      {:ok, %{voicemail: voicemail}} =
+        Multi.new()
+        |> Multi.delete_all(:contacts, where(MatchContact, match_id: ^match_id))
+        |> Multi.delete_all(:timeslots, where(Timeslot, match_id: ^match_id))
+        |> Multi.insert(:voicemail, voicemail)
+        |> Repo.transaction()
+
+      voicemail
```

Comment:
this return probably doesn't match https://github.com/getsince/test3/pull/479/files#diff-a0218a402f6110a957246806c0683922fda46cc95037890c7aae13528effb6a0R371

Reply:
fixed